### ISSUE TITLE
Manager: Logic for choosing modules

### ISSUE THREAD

**Issue by simanadler** [IBM/the-mesh-for-data#93]
The pilot component (M4DApplication controller) contains the logic for choosing the modules to be included in the blueprint.  Currently logic implemented (hard coded) is for example .... (@shlomitk1 keep me honest here)

(1) If the data set protocol/format and the protocol/format requested by the user do not match, then make an implicit copy of the data ,storing it such that it is readable via the protocol/format requested by the user.  
(2) If the governance action required on the data set are not supported by the read module, and it is supported by the implicit copy module ... then make an implicit copy.  Otherwise no need for implicit copy, and read will be done from the source directly.

There could be other rules that an organization might want to implement that would influence how to build the blueprint which we do not currently take into account.  For example ...
(3) If the data set is very large, aim to not use implicit copy.  Prefer direct use of the source.
(4) If performance is important, keep the data as close to the compute as possible.

We need a mechanism (1) for an enterprise to define these rules, (2) M4D should leverage these rules rather than hard coding it in the manager logic

**Comment by shlomitk1** [comment]
@simanadler Current logic supports rule (1) only. There is a strict separation of READ and COPY actions that are not mixed. If no read module supports the required actions, an error will be returned.

**Comment by roee88** [comment]
> There is a strict separation of READ and COPY actions that are not mixed

I thought that COPY policies is applied on top of READ policies in case of a copy. I'm not sure what the right approach is. @simanadler do you have any input here?

> The pilot component (M4DApplication controller) contains the logic for choosing the modules to be included in the blueprint. Currently logic implemented (hard coded) is for example .... (@shlomitk1 keep me honest here)
> 
> (1) If the data set protocol/format and the protocol/format requested by the user do not match, then make an implicit copy of the data ,storing it such that it is readable via the protocol/format requested by the user.
> (2) If the governance action required on the data set are not supported by the read module, and it is supported by the implicit copy module ... then make an implicit copy. Otherwise no need for implicit copy, and read will be done from the source directly.
> 
> There could be other rules that an organization might want to implement that would influence how to build the blueprint which we do not currently take into account. For example ...
> (3) If the data set is very large, aim to not use implicit copy. Prefer direct use of the source.
> (4) If performance is important, keep the data as close to the compute as possible.
> 
> We need a mechanism (1) for an enterprise to define these rules, (2) M4D should leverage these rules rather than hard coding it in the manager logic.
> 
> _Should these rules be handled by the policy manager??_

I think that it should be a combination of metadata of the asset, input in the M4DApplication and policies. We should figure out the details. Forward looking, I see the following factors:

**Application requirements**: 
- Protocol and format
- Currently handled by a copy module unless we have a read module that this is its API

**Required transformations**: 
- The combination of all modules together should create a result that enforces all policies 
- The result must be usable by the business logic of the application. That is, if an application needs access to column X as-is but doing a copy redacts it then some blueprints are no-go. As a side note, allowing listing columns in the `M4DApplication` is important for performance aspects, not it's not enough here.
 
**Performance**: 
- Handled by copy and/or caching modules (when we have those, e.g., using [DLF](https://github.com/IBM/dataset-lifecycle-framework) when they add it)
- Can be explicitly requested in `M4DApplication` ("I want to read asset X but create a copy and keep it constantly updated please")
- Should eventually be auto discovered by asset metadata (e.g., size, source type, location) and collected metrics (e.g., bandwidth)

**Location-based policies**: 
- Is direct access from TARGET to SOURCE allowed?
- Is a copy from SOURCE to TARGET allowed?
- What are the implications on how the data looks like if we make a copy


Your text mostly focuses on performance. We may want to start with something more explicit like extending "copy" to include the same concepts that the mover defines (snapshot, stream, cron) and get these as inputs in the `M4DApplication` ("I want to read asset X but create a copy and keep it constantly updated please") to force a copy even if policies allow read. Later on support implicitly setting these according to metadata and discovered metrics. Same for any other "smartness", start explicit and gradually build the mechanisms to replace with something automatic. 

\+ @froesef

**Comment by simanadler** [comment]
Forward looking, I see the following factors:
> 
> **Application requirements**:
> 
>     * Protocol and format
> 
>     * Currently handled by a copy module unless we have a read module that this is its API
Other application requirements such as:
* Experimentation vs production performance - affecting Read/write latency, Time to first byte (preparation time vs ongoing read/write latency), etc
* Subset of the data required - let's not copy/read data not needed
> 
> 
> **Required transformations**:
> 
>     * The combination of all modules together should create a result that enforces all policies
> 
>     * The result must be usable by the business logic of the application. That is, if an application needs access to column X as-is but doing a copy redacts it then some blueprints are no-go. As a side note, allowing listing columns in the `M4DApplication` is important for performance aspects, not it's not enough here.
> 
> 
> **Performance**:
> 
>     * Handled by copy and/or caching modules (when we have those, e.g., using [DLF](https://github.com/IBM/dataset-lifecycle-framework) when they add it)
> 
>     * Can be explicitly requested in `M4DApplication` ("I want to read asset X but create a copy and keep it constantly updated please")
> 
>     * Should eventually be auto discovered by asset metadata (e.g., size, source type, location) and collected metrics (e.g., bandwidth)
> 
> 
> **Location-based policies**:
> 
>     * Is direct access from TARGET to SOURCE allowed?
> 
>     * Is a copy from SOURCE to TARGET allowed?
> 
>     * What are the implications on how the data looks like if we make a copy
> 
Cluster Capabilities:
* Available compute, bandwidth, storage, etc
* Infrastructure Compliance rating - i.e. appropriate for health data or other special sensitive data?

Geopolitical Influences (really future):
* Avoid clusters in areas affected by floods, earthquakes, political unrest, etc
> 

In short, I think we can categorize the input parameters driving the decisions on how to build the data plane into the following categories:
1. User input
2. Data Set metadata
3. Data Governance policies
4. Infrastructure metadata
5. System level governance policies

@roee88 @froesef @shlomitk1 @ronenkat

**Comment by shlomitk1** [comment]
This is a large issue (epic?) that can be split into sub-issues. 
For the short-term the following sub-issues can be implemented:
- Add column names in m4dapplication (will affect whether apply copy or not: no copy if actions include transformation of required columns, prefer copy if the number of required columns divided by overall column number is less than a certain threshold). 
- Add user preferences (copy always, read directly, ...)
- Cluster rating as a part of metadata? Could be updated automatically in the future collecting metrics, all geopolitical considerations can be translated into ratings as well.

**Comment by ronenkat** [comment]
@shlomitk1
I am not sure if that alone is a good metric. Consider the case where we only read the 1st 10 rows.
User hints (i.e., preferences) can definitely help us.

@simanadler 
I think we also need to consider the dimensions of decision that the manager can make:
1. What modules to deploy - type and implementation
1. What functionality to perform with what type of module
1. Where to deploy each module (in multi cluster settings)

I also suggest we start with a small subset of the input categorizes to have a simple start. I would suggest we start with user input and governance policies.

What is our algorithmic approach for selecting the module? should we build all possible paths and rank them? choose the 1st one we identify?

**Comment by roee88** [comment]
> This is a large issue (epic?) that can be split into sub-issues.

I would say it's more than an epic, it's the vision for the application controller. We would eventually drive multiple epics from it I guess.  

> @shlomitk1
> I am not sure if that alone is a good metric. Consider the case where we only read the 1st 10 rows.
> User hints (i.e., preferences) can definitely help us.

Shlomit described what we can start with and I agree with her about the first two items (also the third but too futuristic to be useful now). Instead of adding verbose user hints like "I need 10 rows" we should wait for a real requirement to drive it and decide how best to achieve it (e.g., how to handle preview). Starting with column selection is a good choice because that's also what is typically requested in data access requests today.

**Comment by simanadler** [comment]
@roee88 @shlomitk1 @ronenkat @froesef @rohithdv Added at the top of the issue description the link to a draft requirements / high level design document

**Comment by roytman** [comment]
@simanadler what status of this issue, can it be closed?
